table/metadata.go

```diff
@@ -0,0 +1,401 @@
+	// to be used for arbitrary metadata. For example, commit.retry.num-retries
+	// is used to control the number of commit retries.
+	Properties() iceberg.Properties
+}
+
+var (
+	ErrInvalidMetadataFormatVersion = errors.New("invalid or missing format-version in table metadata")
+	ErrInvalidMetadata              = errors.New("invalid metadata")
+)
+
+// ParseMetadata parses json metadata provided by the passed in reader,
+// returning an error if one is encountered.
+func ParseMetadata(r io.Reader) (Metadata, error) {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseMetadataBytes(data)
+}
+
+// ParseMetadataString is like [ParseMetadata], but for a string rather than
+// an io.Reader.
+func ParseMetadataString(s string) (Metadata, error) {
+	return ParseMetadataBytes([]byte(s))
+}
+
+// ParseMetadataBytes is like [ParseMetadataString] but for a byte slice.
+func ParseMetadataBytes(b []byte) (Metadata, error) {
+	ver := struct {
+		FormatVersion int `json:"format-version"`
+	}{}
+	if err := json.Unmarshal(b, &ver); err != nil {
+		return nil, err
+	}
+
+	var ret Metadata
+	switch ver.FormatVersion {
+	case 1:
+		ret = &MetadataV1{}
+	case 2:
+		ret = &MetadataV2{}
+	default:
+		return nil, ErrInvalidMetadataFormatVersion
+	}
+
+	return ret, json.Unmarshal(b, ret)
+}
+
+// https://iceberg.apache.org/spec/#iceberg-table-spec
+type commonMetadata struct {
+	FormatVersion      int                     `json:"format-version"`
+	UUID               uuid.UUID               `json:"table-uuid"`
+	Loc                string                  `json:"location"`
+	LastUpdatedMS      int64                   `json:"last-updated-ms"`
+	LastColumnID       int                     `json:"last-column-id"`
+	SchemaList         []*iceberg.Schema       `json:"schemas"`
+	CurrentSchemaID    int                     `json:"current-schema-id"`
+	Specs              []iceberg.PartitionSpec `json:"partition-specs"`
+	DefaultSpecID      int                     `json:"default-spec-id"`
+	LastPartitionID    *int                    `json:"last-partition-id,omitempty"`
+	Props              iceberg.Properties      `json:"properties"`
+	SnapshotList       []Snapshot              `json:"snapshots,omitempty"`
+	CurrentSnapshotID  *int64                  `json:"current-snapshot-id,omitempty"`
+	SnapshotLog        []SnapshotLogEntry      `json:"snapshot-log"`
+	MetadataLog        []MetadataLogEntry      `json:"metadata-log"`
+	SortOrderList      []SortOrder             `json:"sort-orders"`
+	DefaultSortOrderID int                     `json:"default-sort-order-id"`
+	Refs               map[string]SnapshotRef  `json:"refs"`
+}
+
+func (c *commonMetadata) TableUUID() uuid.UUID       { return c.UUID }
+func (c *commonMetadata) Location() string           { return c.Loc }
+func (c *commonMetadata) LastUpdatedMillis() int64   { return c.LastUpdatedMS }
+func (c *commonMetadata) LastColumn() int            { return c.LastColumnID }
```

Comment:
LastColumn() sounds more like a string while the return type is int ?

Reply:
imo this should be named `LastColumnID()`